File: C++/008 Sorting/merge-sort.cpp
#include<bits/stdc++.h>
using namespace std;

void mergeArr(int arr[], int low, int mid, int high){
    vector<int> temp;
    int left = low;
    int right = mid + 1;

    while(left<=mid && right <=high){
        if(arr[left] <= arr[right]){
            temp.push_back(arr[left]);
            left++;
        }
        else{
            temp.push_back(arr[right]);
            right++;
        }
    }
    while(left<=mid){
        temp.push_back(arr[left]);
        left++;

    }
    while(right<=high){
        temp.push_back(arr[right]);
        right++;
        
    }
    for(int i=low;i<=high;i++){
        arr[i] = temp[i-low];
    }
}
void merge_sort(int arr[], int low, int high){
    /*
        Divide and Merge
        Time Complexity - O(n * log base2 n) in all cases
        Space Complexity - O(n)
    */ 
    if(low>=high)  return;
    int mid = (low + high) / 2;
    merge_sort(arr, low, mid);
    merge_sort(arr, mid+1, high);
    mergeArr(arr, low, mid, high);
}
int main()
{
    int arr[] = {3,1,2,4,1,5,2,6,4};
    int n = sizeof(arr) / sizeof(arr[0]);
    int low = 0, high = n - 1;
    merge_sort(arr, low, high);
    for (auto it : arr)
    {
        cout << it << " ";
    }
    return 0;
}

File: JavaScript/002 Sorting/merge-sort.js
/*
    Time complexity: O(nlogn) 

    Reason: At each step, we divide the whole array, for that logn and we assume n steps are taken to get sorted array, so overall time complexity will be nlogn

    Space complexity: O(n)  

    Reason: We are using a temporary array to store elements in sorted order.

    Auxiliary Space Complexity: O(n)
*/

function merge(arr, low, mid, high) {
    let temp = []; // Temporary array
    let left = low; // Starting index of left half
    let right = mid + 1; // Starting index of right half

    // Storing elements in the temporary array in sorted manner
    while (left <= mid && right <= high) {
        if (arr[left] <= arr[right]) {
            temp.push(arr[left]);
            left++;
        } else {
            temp.push(arr[right]);
            right++;
        }
    }

    // If elements on the left half are still left
    while (left <= mid) {
        temp.push(arr[left]);
        left++;
    }

    // If elements on the right half are still left
    while (right <= high) {
        temp.push(arr[right]);
        right++;
    }

    // Transferring all elements from temporary to arr
    for (let i = low; i <= high; i++) {
        arr[i] = temp[i - low];
    }
}

function mergeSort(arr, low, high) {
    if (low >= high) return;
    const mid = Math.floor((low + high) / 2);
    mergeSort(arr, low, mid); // Left half
    mergeSort(arr, mid + 1, high); // Right half
    merge(arr, low, mid, high); // Merging sorted halves
}

const arr = [9, 4, 7, 6, 3, 1, 5];

console.log("Before Sorting Array:");
console.log(arr.join(" "));
mergeSort(arr, 0, arr.length - 1);
console.log("After Sorting Array:");
console.log(arr.join(" "));